### PR TITLE
V3/do not add newline to test body lines

### PR DESCRIPTION
**Pull request by hnakamur**
## what

- Stop adding newline to each line of request or response body in regression tests.
- Add `format` subcommand to `regression_tests` command.
- Support updating `Content-Length` header by running `(cd test; UPDATE_CONTENT_LENGTH=1 ./regression_tests format)`
- Update regression test JSON files:
  - Add newlines to request bodies where needed.
  - Add Content-Length headers.
  - Adjust test casses to pass after above modifications.

## why

- Enable to test edge cases (such as request body lines with newline except the last line). This is important for multipart request bodies.

## references

- https://github.com/owasp-modsecurity/ModSecurity/pull/3479

### CODE REVIEW

**Comment by hnakamur** [comment]
I found JSON was corrupted since some of http_version was empty. I fixed it and force pushed.
However some of tests do not pass right now. I am trying to fix them.

**Comment by hnakamur** [comment]
I fixed tests at aefd4a4 and now all tests pass again.

**Comment by airween** [comment]
Hi @hnakamur,

thank you for this improved PR.

Unfortunately there are two warnings from [cppcheck](https://github.com/owasp-modsecurity/ModSecurity/actions/runs/21342077836/job/61423010127?pr=3486):
```
2026-01-26T00:29:37.2162370Z warning: test/regression/regression.cc,444,style,shadowVariable,Local variable 'test' shadows outer variable
2026-01-26T00:29:37.2943530Z warning: test/regression/regression_test.cc,366,style,functionConst,The member function 'modsecurity_test::RegressionTests::toJSON' can be made a const function. Making this function 'const' should not cause compiler errors. Even though the function can be made const function technically it may not make sense conceptually. Think about your design and the task of the function first - is it a function that must not change object internal state?
```

Also, SonarCloud reported a few new issues (some of them are same as `cppcheck` found):

https://sonarcloud.io/project/issues?id=owasp-modsecurity_ModSecurity&pullRequest=3486&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true

As I mentioned previously, I don't expect to fix all Sonar issues. If the `cppcheck` will be passed, I'll review all Sonar issues and try to mark the "old" issues as accepted.

Thank you again.

**Comment by airween** [comment]
Hi @hnakamur,

finally I was able to start to review this PR, but it's very huge... :smiley:, so I asked Copilot (mainly to review the tests' changes).

It seems like everything is okay, Copilot made one suggestion - would you mind to take a look at that?

And thank you again.

File: test/test-cases/regression/secruleengine.json [hunk 1]
```diff
@@ -1,72 +1,202 @@
...
+    },
+    "request": {
+      "headers": {
+        "Content-Length": "0"
+      },
+      "uri": "",
+      "method": "",
+      "body": [
+        ""
+      ]
```
[the remark this answers is not in the document]
Comment (hnakamur) [reply test/test-cases/regression/secruleengine.json]: Addressed at https://github.com/owasp-modsecurity/ModSecurity/pull/3486/changes/c795b0925df9c669921400c35ffbf02a629e39c3

File: test/test-cases/regression/request-body-parser-multipart-crlf.json [hunk 2]
```diff
@@ -1,67 +1,67 @@
...
+        "1.2\r\n",
+        "1.3\r\n",
+        "-----------------------------69343412719991675451336310646\n",
+        "Content-Disposition: form-data; name=\"b\"\r\n",
+        "\r\n",
+        "2\r\n",
+        "2.1\r\n",
+        "2.2\r\n",
+        "2.3\r\n",
+        "-----------------------------69343412719991675451336310646--\n"
```
[the remark this answers is not in the document]
Comment (hnakamur) [reply test/test-cases/regression/request-body-parser-multipart-crlf.json]: This test expects MULTIPART_STRICT_ERROR because of inconsistency of
CRLF or LF after boundaries.

I made this test clearer at https://github.com/owasp-modsecurity/ModSecurity/pull/3486/changes/f72106d8b3997410371423eaac00f3d7b5e7ae30

File: test/regression/regression.cc [hunk 3]
```diff
@@ -436,6 +437,35 @@ int main(int argc, char **argv)
...
     test.cmd_options(argc, argv);
+
+    if (test.m_format) {
+#ifdef WITH_YAJL
+        std::cout << "start formatting test case JSON files" << std::endl;
+        ModSecurityTest<RegressionTests> test2;
+        test2.cmd_options(argc, argv);
+        test2.load_tests();
+        for (const auto &[name, tests] : test2) {
+            std::ofstream ofs{name};
```
[the remark this answers is not in the document]
Comment (hnakamur) [reply test/regression/regression.cc:448]: I changed to use `std::map` instead of `std::unordered_map` for consistent order of output at https://github.com/owasp-modsecurity/ModSecurity/pull/3486/changes/b8d37a062e4bba38aa4b89f09920ea287bf1a245

File: test/regression/regression_test.h [hunk 4]
```diff
@@ -76,8 +78,34 @@ class RegressionTest {
...
+class RegressionTests {
+ public:
+    static std::unique_ptr<RegressionTests> from_yajl_node(const yajl_val &);
+    void update_content_lengths();
+    std::string toJSON() const;
+
+    std::string filename;
+    std::string name;
+
+    std::vector<RegressionTest> tests;
```
[the remark this answers is not in the document]
Comment (hnakamur) [reply test/regression/regression_test.h]: Resolved at https://github.com/owasp-modsecurity/ModSecurity/pull/3486/changes/edbd4b8a7a5c0c3b3254eaed3f86f8eb5593770e

**Comment by hnakamur** [comment]
@airween Thank you for your comment. I added commits to resolve issues reported by Copilot.
Would you take a look?
Thank you!

**Comment by airween** [comment]
Hi @hnakamur,

> @airween Thank you for your comment. I added commits to resolve issues reported by Copilot. Would you take a look? Thank you!


thank - now there is a new [Sonar issue](https://sonarcloud.io/project/issues?sinceLeakPeriod=true&issueStatuses=OPEN%2CCONFIRMED&pullRequest=3486&id=owasp-modsecurity_ModSecurity&open=AZw2vtwqYaNWWAJeOdNb), could you fix that?

**Comment by hnakamur** [comment]
It turned out the [Sonar issue](https://sonarcloud.io/project/issues?sinceLeakPeriod=true&issueStatuses=OPEN%2CCONFIRMED&pullRequest=3486&id=owasp-modsecurity_ModSecurity&open=AZw4KcMlNIaq5Otw7WWB) is a false positive.
The `RegressionTest::update_content_lengths` method modifies the state of the test, so the loop variable `test` cannot be const.

**Comment by airween** [comment]
Hi @hnakamur,

thank you again for this awesome PR.

There are a lot of changes in tests - I think it's obvious that this can't be reviewed one by one.

For the first look, it seems all tests are structured and organized - which is excellent. I assume you used a tool which produced these files.

If this is true and if it's possible, could you share that tool (may be through a gist)? I would like to reproduce your results just for sure.

Thank you.

File: test/test-cases/regression/action-tnf-base64.json [hunk 5]
```diff
@@ -1,88 +1,92 @@
...
       },
-      "body":[
+      "body": [
         "no need."
       ]
     },
-    "expected":{
-      "debug_log": "t:base64encode: \"dmFsdWUyCg==\""
+    "expected": {
+      "debug_log": "t:base64encode: \"dmFsdWUy\"",
```
Comment (airween) [inline test/test-cases/regression/action-tnf-base64.json:40]: The correct encoded base64 must contain two padding chars, I should keep those.

```suggestion
      "debug_log": "t:base64encode: \"dmFsdWUy==\"",
```

File: test/test-cases/regression/request-body-parser-multipart.json [hunk 6]
```diff
@@ -1,3349 +1,3406 @@
...
+      "body": [
+        "-----------------------------69343412719991675451336310646\r\n",
+        "Content-Disposition: form-data; name=\"a\"\r\n",
+        "\r\n",
+        "1\r\n",
+        "-----------------------------69343412719991675451336310646\r\n",
+        "Content-Disposition: form-data; name=\"b\"\r\n",
+        "\r\n",
+        "2\r\n",
+        "-----------------------------69343412719991675451336310646--"
```
Comment (airween) [inline test/test-cases/regression/request-body-parser-multipart.json:155]: I think you should add trailing EOL chars here.

```suggestion
        "-----------------------------69343412719991675451336310646--\r\n"
```

File: test/test-cases/regression/request-body-parser-multipart-crlf.json [hunk 7]
```diff
@@ -1,67 +1,64 @@
...
-        "SecRuleEngine On",
-        "SecRequestBodyAccess On",
-        "SecRule MULTIPART_STRICT_ERROR \"@eq 1\" \"phase:2,deny,id:500055\"",
-        "SecRule MULTIPART_UNMATCHED_BOUNDARY \"@eq 1\" \"phase:2,deny,id:500056\"",
-        "SecRule REQBODY_PROCESSOR_ERROR \"@eq 1\" \"phase:2,deny,id:500057\"",
-        "SecRule ARGS_POST \"@eq 1231\" \"phase:2,deny,id:500067\""
+    "rules": [
+      "SecRuleEngine On",
+      "SecRequestBodyAccess On",
+      "SecRule MULTIPART_STRICT_ERROR \"@eq 1\" \"phase:2,deny,id:500055\""
```
Comment (airween) [inline test/test-cases/regression/request-body-parser-multipart-crlf.json:61]: I think here it would be nice to put the log the reason of the `MULTIPART_STRICT_ERROR`, with the `msg` action:

```suggestion
      "SecRule MULTIPART_STRICT_ERROR \"@eq 1\" \"phase:2,deny,id:500055,msg:'Multipart request body failed strict validation:PE %{REQBODY_PROCESSOR_ERROR},BQ %{MULTIPART_BOUNDARY_QUOTED},BW %{MULTIPART_BOUNDARY_WHITESPACE},DB %{MULTIPART_DATA_BEFORE},DA %{MULTIPART_DATA_AFTER},HF %{MULTIPART_HEADER_FOLDING},LF %{MULTIPART_LF_LINE},SM %{MULTIPART_MISSING_SEMICOLON},IQ %{MULTIPART_INVALID_QUOTING},IP %{MULTIPART_INVALID_PART},IH %{MULTIPART_INVALID_HEADER_FOLDING},FL %{MULTIPART_FILE_LIMIT_EXCEEDED}'\""
```
and add `error_log` section under the `expected` keys like this:
```
    "error_log": "msg \"Multipart request body failed strict validation:PE 0,BQ 0,BW 0,DB 0,DA 0,HF 0,LF 1,SM 0,IQ 0,IP 0,IH 0,FL \""
```
What do you think?

**Review by airween — CHANGES_REQUESTED** [review]

File: test/test-cases/regression/action-tnf-base64.json (hunk 5 above)
Comment (hnakamur) [reply test/test-cases/regression/action-tnf-base64.json:40]: No padding is needed here since I removed the last LF from the request body and the modified body length is 27 which is divisible by 3.
I removed the last LF since I think typical request bodies of `Content-Type: application/x-www-form-urlencoded` do not end with LF.

File: test/test-cases/regression/request-body-parser-multipart.json (hunk 6 above)
Comment (hnakamur) [reply test/test-cases/regression/request-body-parser-multipart.json:155]: Since the test case title says "multipart parser (no final CRLF)", I think it is intentional that the final line does not end with CRLF here.

File: test/test-cases/regression/request-body-parser-multipart-crlf.json (hunk 7 above)
Comment (hnakamur) [reply test/test-cases/regression/request-body-parser-multipart-crlf.json:61]: Nice idea! I added the `msg` action to tests cases whose status is 403 and the cause of `deny` is `SecRule MULTIPART_STRICT_ERROR "@eq 1"`

File: test/test-cases/regression/request-body-parser-multipart.json (hunk 6 above)
Comment (airween) [reply test/test-cases/regression/request-body-parser-multipart.json:155]: Ah, yes, sorry, that makes sense. Thanks.

File: test/test-cases/regression/request-body-parser-multipart-crlf.json (hunk 7 above)
Comment (airween) [reply test/test-cases/regression/request-body-parser-multipart-crlf.json:61]: It seems like you implemented this in 07c487bcfc6ecff51e6d2637d73853a1c27eb846, thanks.

File: test/test-cases/regression/action-tnf-base64.json (hunk 5 above)
Comment (airween) [reply test/test-cases/regression/action-tnf-base64.json:40]: I see, sorry. You're right.

**Review by airween — APPROVED** [review]

**Comment by airween** [comment]
Hi @hnakamur,

thank you so much for your effort and your patience. I see this is a huge PR and I'm sure you spent a lot of time with this.

I don't know if you seen [this](https://github.com/owasp-modsecurity/ModSecurity/pull/3486#issuecomment-3884691935) comment, it's not important now - I solved the comparison in other way.

But for the future: now - as I wrote - you organized and cleaned all the test cases, it would be nice to have to keep this strict format. So if you have more patience, and have some idea how could we implement a GH CI job where we can check the syntax of all tests, and add some tool to maintain the tests for other developer, please share that with us.

And thank you again for this huge PR.

**Comment by hnakamur** [comment]
Thank your for your thourough review!

> I don't know if you seen [this](https://github.com/owasp-modsecurity/ModSecurity/pull/3486#issuecomment-3884691935) comment, it's not important now - I solved the comparison in other way.

Sorry I had missed that comment.

Yes, I added the `format` subcommand to `regression_tests` command at https://github.com/owasp-modsecurity/ModSecurity/pull/3486/changes/b1c59b72504c23e51b95c84238c11e21f22b394a.
I also added support for updating the `Content-Length` request headers at https://github.com/owasp-modsecurity/ModSecurity/pull/3486/changes/37e1f133d72b75bbe2861d5d31fbc9495818061a

To format regression test case json files, run the following command:

```
(cd test && ./regression_tests format)
```
The format subcommand takes filename arguments and formats only the specified files if filename arguments are present.

To update the `Content-Length` request headers, run the following command:
```
(cd test && UPDATE_CONTENT_LENGTH=1 ./regression_tests format)
```
The above command adds or updates the `Content-Length` request header when the test case does not have `Transfer-Encoding: chunked` in request headers.

Maybe we can add the latter command to the git pre-commit hook, or we can just add the explanation to README
since we may want a test case whose `Content-Length` value is intentionally different from the actual length.

**Comment by airween** [comment]
> Thank your for your thourough review!

yw!
 
> Yes, I added the `format` subcommand to `regression_tests` command at [b1c59b7](https://github.com/owasp-modsecurity/ModSecurity/commit/b1c59b72504c23e51b95c84238c11e21f22b394a). I also added support for updating the `Content-Length` request headers at [37e1f13](https://github.com/owasp-modsecurity/ModSecurity/commit/37e1f133d72b75bbe2861d5d31fbc9495818061a)

sorry, I didn't realize that.
 
> To format regression test case json files, run the following command:
> 
> ```
> (cd test && ./regression_tests format)
> ```
> 
> The format subcommand takes filename arguments and formats only the specified files if filename arguments are present.
> 
> To update the `Content-Length` request headers, run the following command:
> 
> ```
> (cd test && UPDATE_CONTENT_LENGTH=1 ./regression_tests format)
> ```
> 
> The above command adds or updates the `Content-Length` request header when the test case does not have `Transfer-Encoding: chunked` in request headers.

That's awesome!

I'm going to update the README, especially [this](https://github.com/owasp-modsecurity/ModSecurity?tab=readme-ov-file#testing-your-patch) section, and add this tool.
 
> Maybe we can add the latter command to the git pre-commit hook, or we can just add the explanation to README since we may want a test case whose `Content-Length` value is intentionally different from the actual length.

Yes, that's also a good idea!
